proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+      setSchedContext (fromJust scPtrOptDonated) (scReply_update (\<lambda>_. Just replyPtr) scDonated);
+      schedContextDonate (fromJust scPtrOptDonated) calleePtr
+    od
+   \<lbrace>\<lambda>_ s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  apply (wp hoare_drop_imp[where R="\<lambda>_ s. ko_at' _ (fromJust scPtrOptDonated) s"] hoare_vcg_if_lift
+            hoare_vcg_all_lift hoare_vcg_const_imp_lift hoare_vcg_imp_lift' setReply_obj_at')
+  apply (auto simp: list_refs_of_replies'_def list_refs_of_reply'_def
+                    opt_map_Some_def obj_at'_def projectKO_eq
+              elim: delta_sym_refs split: if_splits)
+  done
+
+lemma replyPush_sym_refs[wp]:
+  "replyPush callerPtr calleePtr replyPtr canDonate \<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  apply (simp only: replyPush_def)
+  apply (rule replyPush_sym_refs_helper hoare_fun_app_wp hoare_when_wp hoare_seq_ext hoare_K_bind)+
+                      apply (wpsimp wp: threadGet_wp hoare_vcg_all_lift hoare_vcg_if_lift
+                                        hoare_vcg_imp_lift' hoare_vcg_disj_lift gts_wp'
+                                        setReply_obj_at' haskell_assert_wp
+                                  simp: setReplyTCB_def getReplyTCB_def)+
+  apply (fastforce simp: comp_def obj_at'_def projectKO_eq projectKO_tcb)
+  done
+
+lemma replyPush_if_live_then_nonz_cap':
+  "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s \<and> valid_objs' s \<and> valid_idle' s \<and> st_tcb_at' active' callerPtr s \<and>
+        ex_nonz_cap_to' replyPtr s \<and> ex_nonz_cap_to' callerPtr s \<and> ex_nonz_cap_to' calleePtr s\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  apply (rule monadic_rewrite_refine_valid[OF monadic_rewrite_replyPush, where P''=\<top>, simplified])
+  apply (wp schedContextDonate_if_live_then_nonz_cap')
+        apply (rule_tac Q="\<lambda>_ s. (\<forall>scp. scPtrOptDonated = Some scp \<longrightarrow> ex_nonz_cap_to' scp s) \<and>
+                                 ex_nonz_cap_to' calleePtr s \<and>
+                                 valid_objs' s \<and> if_live_then_nonz_cap' s"
+                     in hoare_strengthen_post[rotated], clarsimp)
+        apply (wp hoare_vcg_all_lift hoare_vcg_const_imp_lift)
+        apply (simp (no_asm_use) add: replyPush'_def getReplyTCB_def split del: if_split
+               | wp hoare_vcg_all_lift hoare_vcg_disj_lift hoare_vcg_imp_lift'
+                    hoare_vcg_if_lift setReply_obj_at' setReplyTCB_obj_at'
+               | rule threadGet_wp)+
+  apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_tcb projectKO_sc projectKO_reply)
+  apply (rule context_conjI; clarsimp simp: sym_refs_asrt_def)
+   apply (erule if_live_then_nonz_capE')
+   apply (drule_tac ko=obj in sym_refs_ko_atD'[rotated])
+    apply (auto simp: refs_of_rev' live_sc'_def idle_tcb'_def valid_idle'_def
```

Comment:
Did we ever agree on a standard when using auto in the middle of a proof? Was it "subgoal by" that we wanted? (And check to see if a fastforce or something else is enough here, if you haven't already)

Reply:
I don't think that we have a strong rule on this, but the preference is to avoid `auto` in the middle of proofs. In this case it's a bit more complicated since the auto is solving two goals, and we also don't want to just copy and paste it if we can avoid it.

In my mind, the optimal but possibly hard than it's worth solution would be to somehow handle all three of the final subgoals together, in one terminating `auto`. The other options that I'm equally happy with are leaving it as is, or using `prefer` to handle the `fastforce` first, before finishing with the `auto`.

@mbrcknl @Xaphiosis @lsf37 @mktnk3 Do any of you have an opinion?

Reply:
Resolved by changing the auto instance into two separate fastforces